### PR TITLE
Catch all exceptions on ANCM/IIS border

### PR DESCRIPTION
**Pull request by pakrym**
new operator might throw, catch all exceptions before they reach IIS.

### CODE REVIEW

File: src/AspNetCoreModuleV2/CommonLib/exceptions.h [hunk 1]
```diff
@@ -0,0 +1,35 @@
+#pragma once
```
Comment (jkotalik) [inline src/AspNetCoreModuleV2/CommonLib/exceptions.h]: copyright header.

File: src/AspNetCoreModuleV2/CommonLib/exceptions.h [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+{
+    OutputDebugStringA("Unhandled exception:");
+    OutputDebugStringA(exception.what());
+}
+
+inline __declspec(noinline) HRESULT CaughtExceptionHResult()
+{
+    try
+    {
+        throw;
```
Comment (jkotalik) [inline src/AspNetCoreModuleV2/CommonLib/exceptions.h]: Tests for this function.

File: src/AspNetCoreModuleV2/AspNetCore/Inc/proxymodule.h [hunk 3]
```diff
@@ -61,4 +61,4 @@ class ASPNET_CORE_PROXY_MODULE_FACTORY : public IHttpModuleFactory
 
     VOID
     Terminate();
-};
\ No newline at end of file
+};
```
Comment (jkotalik) [inline src/AspNetCoreModuleV2/AspNetCore/Inc/proxymodule.h]: revert

File: src/AspNetCoreModuleV2/AspNetCore/src/proxymodule.cxx [hunk 4]
```diff
@@ -83,96 +84,104 @@ ASPNET_CORE_PROXY_MODULE::OnExecuteRequestHandler(
     REQUEST_NOTIFICATION_STATUS retVal = RQ_NOTIFICATION_CONTINUE;
     IAPPLICATION* pApplication = NULL;
     STACK_STRU(struFileName, 256);
-
-    if (g_fInShutdown)
+    try
```
Comment (jkotalik) [inline src/AspNetCoreModuleV2/AspNetCore/src/proxymodule.cxx]: Why not wrap the entire function in try catch?

Comment (pakrym) [reply src/AspNetCoreModuleV2/AspNetCore/src/proxymodule.cxx]: returns REQUEST_NOTIFICATION_STATUS and has to set 500 response code.

File: src/AspNetCoreModuleV2/CommonLib/exceptions.h (hunk 2 above)
Comment (pakrym) [reply src/AspNetCoreModuleV2/CommonLib/exceptions.h]: Good point.

File: test/CommonLibTests/exception_handler_tests.cpp [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    }
+    catch(...)
+    {
+        hr = CaughtExceptionHResult();
+    }
+
+    EXPECT_EQ(E_INVALIDARG, hr);
+}
+
+TEST(CaughtExceptionHResult, ReturnsERROR_UNHANDLED_EXCEPTIONForOtherExceptions)
```
Comment (jkotalik) [inline test/CommonLibTests/exception_handler_tests.cpp]: Cleanup these names.  ReturnsErrorUnhandledExceptionForOtherExceptions

**Comment by pakrym** [comment]
No perf regression:
```
[10:35:13.768] RequestsPerSecond:           214,168
[10:35:13.768] Max CPU (%):                 8623
[10:35:13.769] WorkingSet (MB):             404
[10:35:13.769] Avg. Latency (ms):           1.2
[10:35:13.769] Startup (ms):                511
[10:35:13.770] First Request (ms):          424.8
[10:35:13.770] Latency (ms):                0.9
[10:35:13.770] Total Requests:              3,233,828
[10:35:13.770] Duration: (ms)               15,100
[10:35:13.770] Socket Errors:               0
[10:35:13.771] Bad Responses:               0
```

File: src/AspNetCoreModuleV2/AspNetCore/src/proxymodule.cxx [hunk 6]
```diff
@@ -204,7 +214,14 @@ ASPNET_CORE_PROXY_MODULE::OnAsyncCompletion(
     IHttpCompletionInfo *   pCompletionInfo
 )
 {
-    return m_pHandler->OnAsyncCompletion(
-        pCompletionInfo->GetCompletionBytes(),
-        pCompletionInfo->GetCompletionStatus());
+    try
+    {
```
[the remark this answers is not in the document]
Comment (pakrym) [reply src/AspNetCoreModuleV2/AspNetCore/src/proxymodule.cxx]: Anything inside m_pHandler->OnAsyncCompletion

[the remark this answers is not in the document]
Comment (jkotalik) [reply src/AspNetCoreModuleV2/AspNetCore/src/proxymodule.cxx]: Manually try throwing many exceptions in the modules and see what happens? You can write a test by creating your own module but is probably too much work for this change.

**Comment by jkotalik** [comment]
appveyor failed for a flaky test.

**Comment by pakrym** [comment]
@dotnet-bot test this please

File: src/AspNetCoreModuleV2/CommonLib/exceptions.h [hunk 7]
```diff
@@ -0,0 +1,38 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+#pragma once
+inline VOID ReportException(std::exception& exception)
+{
```
Comment (pan-wang) [inline src/AspNetCoreModuleV2/CommonLib/exceptions.h]: I think we should only report exception for debug build and check whether debug flag enabled as the cost could be very high for release scenario. Have you tested the behavior for the Remarks scenario in https://msdn.microsoft.com/en-us/library/windows/desktop/aa363362(v=vs.85).aspx

Comment (jkotalik) [reply src/AspNetCoreModuleV2/CommonLib/exceptions.h]: The debug flag is undiscoverable and hard to set. I agree that we shouldn't call OutputDebugStringA for every request, but this is only in the failure path.

Comment (jkotalik) [reply src/AspNetCoreModuleV2/CommonLib/exceptions.h]: Maybe make it on by default would be better.

**Comment by jkotalik** [comment]
@pan-wang can you list the modifications you want for this PR?

**Comment by pan-wang** [comment]
Basically I want enable nothrownew in compilation so that we would not hit memory leaks caused by new failure.  Otherwise, we have to deeply scan all functions to wrap the NEW calls and make sure there is no memory leak.

**Comment by pakrym** [comment]
There are better ways to address this problem then compiling against obsolete non standard compliant library. 

We have other places that might throw, and catching and reporting is a nice thing to do.

**Comment by pakrym** [comment]
The other thing is, CLR would crash anyway in OOM situations.

**Comment by pan-wang** [comment]
No. for web server, reliability is a key. You don't want to crash the process even in memory critical situation. Current try-catch, can leak memory and make things even worse. I was not convinced by CLR would crash as I never see it. Current try-catch code opens doors for lots of unknowns. Please following IIS coding guideline which has its good reason on avoiding try-catch.  If you really like, you can start a conversation with Bila or Drago.

**Comment by pakrym** [comment]
CLR and ASP.NET Core are not hardened against OOM and don't catch OutOfMemoryExceptions anywhere so application would almost certainly crash under low memory.

As for leaking memory in when throwing exceptions: there are modern C++ primitives that prevent such leaks from happening in any situation and simplify memory management (take a look at unique_ptr, shared_ptr in C++11). We should certainly look into and start using them. Unknowns are part of the progress.

I don't think that referring to the IIS internal guidance and it's authors is a productive way to discuss technical problems. With all due respect to authors of the guideline it is 20 years old and many things have changed in C++ world including patterns for error handling and exception management.

Additionally we are not trying to contribute code to IIS and have product specifics that are not covered by the said guidance.

1. Much shorter agile release cycle.
2. Being open source and accessible to external contributors.
3. Being easily understood by new developers on the project.

We are free to select the guidance that fits our product the best.

**Comment by pakrym** [comment]
AppVeyor failure: https://github.com/aspnet/IISIntegration/issues/939

**Pull request merged**